kernel/src/expressions/transforms.rs

```diff
@@ -42,6 +42,20 @@ pub trait ExpressionTransform<'a> {
         self.recurse_into_expr_struct(fields)
     }
 
+    /// Called for each [`OpaqueExpression`] encountered during the traversal. Implementations can
```

Comment:
There's quite a lot of duplicate code in this file. Would it make sense to define an:
```rust
enum ExprOrPred {
  Expr(Expression)
  Pred(Predicate)
}
```

And unify across that type, and then the entry would be:
```rust
fn transform_expr_or_pred(&mut self, expr_or_pred: &'a ExprOrPred) -> Option<Cow<'a, ExprOrPred>>
```

I can spec it out a bit more if you think that might be helpful, or not if there's an obvious reason that doesn't work that I'm not thinking of